<a href="https://colab.research.google.com/github/ameliagarciarond22-source/Modelos-Estoc-sticos-/blob/main/Matriz_Fundamental.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Serpientes y escaleras

Sea:

$$
E_i
$$

el número esperado de tiradas para terminar el juego comenzando en la casilla $i$.

La ecuación de esperanza es:

$$
E_i
=
1
+
\frac{1}{6}
\sum_{k=1}^{6}
E_{f(i,k)}
$$

In [2]:
from sympy import *
import sympy as sp

# Escaleras y serpientes

especiales = {
    3: 11,
    15: 19,
    13: 4,
    17: 9
}

In [4]:
def mover(posicion, dado):

    nueva = posicion + dado

    # Si se pasa del final
    if nueva > 20:
        nueva = posicion

    # Aplicar serpiente/escalera
    nueva = especiales.get(nueva, nueva)

    return nueva

In [5]:
## Construcción del sistema de ecuaciones
# Variables simbólicas

E = symbols('E1:21')

ecuaciones = []

In [6]:
# Construcción de ecuaciones

for i in range(1,20):

    ecuacion = Eq(
        E[i-1],
        1 + sum(E[mover(i,d)-1] for d in range(1,7))/6
    )

    ecuaciones.append(ecuacion)

# Estado final absorbente
ecuaciones.append(Eq(E[19],0))

In [7]:
# Resolver sistema

solucion = solve(ecuaciones, E)

print(float(solucion[E[0]]))

11.561999503019546


In [8]:
# Simulación Monte Carlo
import random
import numpy as np

def jugar():

    posicion = 1
    tiros = 0

    while posicion != 20:

        dado = random.randint(1,6)

        tiros += 1

        nueva = posicion + dado

        if nueva > 20:
            nueva = posicion

        nueva = especiales.get(nueva,nueva)

        posicion = nueva

    return tiros

In [9]:
N = 100000

resultados = [jugar() for _ in range(N)]

print(np.mean(resultados))

11.52229


#Ratón y comida

Definimos:

$$
p_i
=
\text{probabilidad de llegar a comida iniciando en } i
$$

Estados absorbentes:

$$
p_7 = 1
$$

$$
p_8 = 0
$$

In [10]:
adyacencia = {
    0:[1,2],
    1:[0,7,3],
    2:[0,3,8],
    3:[1,2,4,5],
    4:[3,7,6],
    5:[3,8,6],
    6:[4,5],
    7:[],
    8:[]
}

In [12]:
p = symbols('p0:9')

ecuaciones = []

for estado in [0,1,2,3,4,5,6]:

    vecinos = adyacencia[estado]

    ecuaciones.append(
        Eq(
            p[estado],
            sum(p[v] for v in vecinos)/len(vecinos)
        )
    )

ecuaciones.append(Eq(p[7],1))
ecuaciones.append(Eq(p[8],0))

In [13]:
sol = solve(ecuaciones,p)

print(sol[p[0]])
print(float(sol[p[0]]))

1/2
0.5


In [14]:
# Simulación Monte Carlo
def raton():

    estado = 0

    while estado not in [7,8]:

        vecinos = adyacencia[estado]

        estado = random.choice(vecinos)

    return 1 if estado == 7 else 0

In [15]:
N = 100000

experimentos = [raton() for _ in range(N)]

print(np.mean(experimentos))

0.50157


# Conclusiones

- El juego de serpientes y escaleras requiere aproximadamente 11.56 tiradas.
- La probabilidad de que el ratón alcance la comida es aproximadamente 0.5.
- Las simulaciones validan los resultados analíticos.